Maintenance/python package

mcw92 opened the pull request:
This PR restructures the whole codebase in form of a Python package. The following things have changed:

- Add pyproject.toml file.
- Add proper .gitignore file.
- Add pre-commit hooks (formatting and linting with Ruff, type checking with MyPy etc.).
- Add automatic test workflow upon push including code coverage.
- Restructure and refactor code into the following modules:
  - `rf_parallel.py`: Actual distributed random forest functionality
  - `synthetic_classification_data.py`: Synthetic classification data generation functionality
  - utils/
    - `__init__.py`: General utility functions
    - `plot.py`: Utility functions for plotting and visualizing results
    - `train.py`: Utility functions for serial and parallel training on balanced and imbalanced synthetic classification data
- Move example bash and Python scripts to separate `scripts/` folder outside of the package.

fluegelk commented on specialcouscous/utils/train.py:
I have also considered if `utils` is the correct place for `train.py` or if it would make more sense in the top level `specialcouscous` package as the training is an essential part of our package and not just utility

mcw92 replied:
I tried to refactor the train functions as much as I can; however, I found it kind of difficult. I moved some code to additional helper functions in the same file. This increased the overall number of lines of the module and also did not really reduce the number of lines of each function. They should be a bit easier to understand now. Maybe you have a better idea how to condense the code.
In addition, I moved the train module one package structure level up.

fluegelk replied:
I would say this is good enough for now, most functions have been reduced to ~100 lines and most of the remaining lenght is due to black's lengthy (in terms of lines of code) code formatting.

mcw92 commented on README.md:
```diff
@@ -1,68 +1,86 @@
+[![](https://img.shields.io/badge/Contact-marie.weiel%40kit.edu-blue)](mailto:marie.weiel@kit.edu)
+![](./coverage.svg)
+
+``SpecialCouscous`` is a Python package for large-scale distributed training and inference of random forests on
```
I changed all occurrences of the package name to ``specialcouscous``.

mcw92 commented on README.md:
```diff
@@ -1,68 +1,86 @@
+## Installation
+- You can install ``SpecialCouscous`` directly from the main branch.
+Pull and run: ``pip install .``
+- If you want to **contribute** to ``SpecialCousCous`` as a developer, you need to install the required dependencies
```
I changed all occurrences of the package name to ``specialcouscous``.

mcw92 commented on .gitignore:
```diff
@@ -0,0 +1,165 @@
+#.idea/
+
+*.pickle
+results/
```
I actually don't remember exactly. I think I used the ``*.pt`` file extension to save model checkpoints. I changed this to ``*.pickle** in accordance with the [official Python3 documentation](https://docs.python.org/3.11/library/pickle.html#examples). The ``*.out`` files probably resulted from me redirecting all stdout to files with that extension. This is no longer needed as we have a logger now.

mcw92 commented on scripts/scaling_experiments/single_node/n7_m3_baseline.sh:
```diff
@@ -1,32 +1,28 @@
-
 export OMP_NUM_THREADS=${SLURM_CPUS_PER_TASK}
-export PYDIR=${BASE_DIR}/special-couscous/specialcouscous
+export PYDIR=/hkfs/work/workspace/scratch/ku4408-special-couscous/special-couscous/specialcouscous
```
No, that's probably because I had two versions of these scripts where one of them used this and the other for whatever reason not. I will change it back to re-include ``${BASE_DIR}`` as I think it's actually useful and clearer.

mcw92 commented on scripts/examples/breaking_iid_demo.sh:
```diff
@@ -0,0 +1,32 @@
+N_TREES=1000  # Number of trees
+
+EXPERIMENT_ID="breaking_iid_demo"
+SCRIPT="../py/rf_training_breaking_iid.py"
```
I fixed the paths hopefully everywhere in the bash scripts.

mcw92 commented on scripts/scaling_experiments/strong_scaling/n6_m4/baseline.sh:
```diff
@@ -0,0 +1,28 @@
+
+N_SAMPLES=1000000
+N_FEATURES=10000
+N_TREES=1600  # We estimated 1500 trees should be trainable in serial in 3 d and chose the next smaller number evenly divisble by 64 (scaling exps)
```
Fixed the comment.

mcw92 commented on scripts/scaling_experiments/strong_scaling/n6_m4/analyze_acc_serial.sh:
```diff
@@ -0,0 +1,28 @@
+
+N_SAMPLES=1000000
+N_FEATURES=10000
+N_TREES=76  # We estimated 1500 trees should be trainable in serial in 3 d and chose the next smaller number evenly divisble by 64 (scaling exps)
```
With the restructuring of the code, this script was no longer needed so I removed it completely.

mcw92 commented on scripts/scaling_experiments/strong_scaling/n6_m4/analyze_acc_parallel.sh:
```diff
@@ -0,0 +1,31 @@
+mkdir ${RESDIR}
+cd ${RESDIR}
+
+srun python -u ${PYDIR}/${SCRIPT} --n_samples ${N_SAMPLES} --n_features ${N_FEATURES} --n_trees ${N_TREES} --output_dir ${RESDIR} --output_label ${SLURM_JOB_ID} --detailed_evaluation
```
I fixed the paths and module names to represent the new structure.

mcw92 commented on scripts/scaling_experiments/strong_scaling/n6_m4/analyze_acc_parallel.sh:
```diff
@@ -0,0 +1,31 @@
+
+N_SAMPLES=1000000
+N_FEATURES=10000
+N_TREES=76  # We estimated 1500 trees should be trainable in serial in 3 d and chose the closest number evenly divisble by 64 as a baseline for scaling exps.
```
I removed the comment.

mcw92 commented on scripts/scaling_experiments/strong_scaling/n7_m3/analyze_acc_serial.sh:
```diff
@@ -0,0 +1,28 @@
+
+N_SAMPLES=10000000
+N_FEATURES=1000
+N_TREES=76  # We estimated 450 trees to be trainable in serial in 3 days and chose the next smaller number evenly divisible by 64 (scaling experiments).
```
With the restructuring of the code, this script was no longer needed (already covered by baseline script) so I removed it completely.

mcw92 commented on scripts/scaling_experiments/strong_scaling/n7_m3/analyze_acc_parallel.sh:
```diff
@@ -0,0 +1,31 @@
+
+N_SAMPLES=10000000
+N_FEATURES=1000
+N_TREES=76  # We estimated 450 trees should be trainable in serial in 3 d and chose the closest number evenly divisble by 64 as a baseline for scaling exps.
```
I removed the comment.

mcw92 commented on scripts/scaling_experiments/weak_scaling/n6_m4/baseline.sh:
```diff
@@ -0,0 +1,28 @@
+
+N_SAMPLES=1000000
+N_FEATURES=10000
+N_TREES=1600  # We estimated 1500 trees should be trainable in serial in 3 d and chose the next smaller number evenly divisble by 64 (scaling exps)
```
Fixed the comment.

fluegelk commented on scripts/scaling_experiments/single_node/n6_m4_baseline.sh:
```diff
@@ -1,28 +1,36 @@
+SCRIPT="scripts/examples/rf_serial_synthetic.py"  # Set script name.
 
+# Create directory to save results to.
 RESDIR=/hkfs/work/workspace/scratch/ku4408-special-couscous/results/single_node_experiments/job_${SLURM_JOB_ID}_n6_m4_baseline/
```
```suggestion
RESDIR=${{BASE_DIR}}/results/single_node_experiments/job_${SLURM_JOB_ID}_n6_m4_baseline/
```

fluegelk commented on scripts/scaling_experiments/strong_scaling/n6_m4/analyze_acc_parallel.sh:
```diff
@@ -1,31 +1,36 @@
-mkdir ${RESDIR}
-cd ${RESDIR}
+# Create directory to save results to.
+RESDIR=/hkfs/work/workspace/scratch/ku4408-special-couscous/results/single_node_experiments/job_${SLURM_JOB_ID}_n6_m4_baseline/
```
```suggestion
RESDIR=${{BASE_DIR}}/results/single_node_experiments/job_${SLURM_JOB_ID}_n6_m4_baseline/
```

fluegelk commented on scripts/scaling_experiments/strong_scaling/n6_m4/baseline.sh:
```diff
@@ -1,28 +1,36 @@
+SCRIPT="scripts/examples/rf_serial_synthetic.py"  # Set script name.
 
+# Create directory to save results to.
 RESDIR=/hkfs/work/workspace/scratch/ku4408-special-couscous/results/single_node_experiments/job_${SLURM_JOB_ID}_n6_m4_baseline/
```
```suggestion
RESDIR=${{BASE_DIR}}/results/single_node_experiments/job_${SLURM_JOB_ID}_n6_m4_baseline/
```

fluegelk commented on scripts/scaling_experiments/strong_scaling/n7_m3/analyze_acc_parallel.sh:
```diff
@@ -1,31 +1,36 @@
-mkdir ${RESDIR}
-cd ${RESDIR}
+# Create directory to save results to.
+RESDIR=/hkfs/work/workspace/scratch/ku4408-special-couscous/results/single_node_experiments/job_${SLURM_JOB_ID}_n6_m4_baseline/
```
```suggestion
RESDIR=${{BASE_DIR}}/results/single_node_experiments/job_${SLURM_JOB_ID}_n6_m4_baseline/
```

fluegelk commented on scripts/scaling_experiments/weak_scaling/n6_m4/baseline.sh:
```diff
@@ -5,24 +5,32 @@
+SCRIPT="scripts/examples/rf_serial_synthetic.py"  # Set script name.
 
+# Create directory to save results to.
 RESDIR=/hkfs/work/workspace/scratch/ku4408-special-couscous/results/single_node_experiments/job_${SLURM_JOB_ID}_n6_m4_baseline/
```
```suggestion
RESDIR=${{BASE_DIR}}/results/single_node_experiments/job_${SLURM_JOB_ID}_n6_m4_baseline/
```

fluegelk commented on scripts/scaling_experiments/weak_scaling/n7_m3/baseline.sh:
```diff
@@ -1,28 +1,36 @@
-mkdir ${RESDIR}
-cd ${RESDIR}
+# Create directory to save results to.
+RESDIR=/hkfs/work/workspace/scratch/ku4408-special-couscous/results/single_node_experiments/job_${SLURM_JOB_ID}_n6_m4_baseline/
```
```suggestion
RESDIR=${{BASE_DIR}}/results/single_node_experiments/job_${SLURM_JOB_ID}_n6_m4_baseline/
```

mcw92 commented on scripts/examples/rf_parallel_synthetic.py:
I refactored the arg parsing into a separate utility function.

mcw92 commented on scripts/scaling_experiments/strong_scaling/n6_m4/generate_job_scripts.py:
I merged the `generate_job_scripts.py` scripts for each strong and weak scaling and added an optional `submit` flag (which defaults to False). If `submit` is True, the scripts will also be submitted to the cluster; if it is False, the scripts will just be generated.

mcw92 commented on scripts/scaling_experiments/single_node/n6_m4_baseline.sh:
I condensed everything into a single single-node baseline script. Using these script with different numbers of samples, numbers of features, and numbers of trees enables one to reproduce all scripts from before. I added explaining comments in the script.

mcw92 commented on specialcouscous/utils/__init__.py:
```diff
@@ -0,0 +1,439 @@
+    return df
+
+
+def time_to_seconds(time_str: str) -> Union[float, None]:
```
I moved it to `utils/slurm.py` as it is more related to processing SLURM output files than to plotting even though it is currently only used in `utils/plot.py`/

mcw92 commented on specialcouscous/utils/__init__.py:
I introduced a couple of new submodules in `utils/` to structure the code more. Putting all the functions into `__init__.py` was an Oskar-style move :D. Only the logger, the argparsing, and the helper function to determine the problem size remain in the `__init__.py` as I did not want to introduce a new submodule for single, comparably small functions.

mcw92 commented:
I tried to address all your comments. However, I would appreciate your feedback in particular on the `train.py` module which I tried to refactor but it's still not really clean.

I re-opened all conversations that are still relevant to this review and hope this helps with keeping the overview of the changes that I made.

You can probably ignore all comments regarding the bash scripts as I deleted most of them from the repository. Now, there is only one single-node baseline script left, which can be used as a baseline for all three experiment series, i.e., single-node capacity and strong + weak scaling. I added explaining comments to the bash script to make this clear. Exemplary scripts for strong and weak scaling, showing how to run distributed training, are contained in the respective job script generation Python scripts `generate_strong_scaling_job_scripts.py` and `generate_weak_scaling_job_scripts.py` (in the `script_content` variable).

fluegelk reviewed (commented):
I went through all your changes, most of the issues have been fixed and I marked the corresponding conversations as resolved. I found some minor things (python version in `pyproject.toml`, log "header" in tests) and would consider also merging the strong and weak scaling experimental scripts as there is still a lot of overlap there.

mcw92 commented on pyproject.toml:
```diff
@@ -0,0 +1,152 @@
+indent-width = 4
+
+# Assume Python 3.9.
+target-version = "py39"
```
Nope, that was a mistake. I bumped the version to 3.10 (because I wanted to have cleaner type hints) and forgot this line. Will fix.

mcw92 commented on scripts/experiments/generate_strong_scaling_job_scripts.py:
```diff
@@ -0,0 +1,110 @@
+        script_content = f"""#!/bin/bash
+#SBATCH --job-name={job_name}  # Job name
+#SBATCH --partition=cpuonly    # Queue for resource allocation
+#SBATCH --time={time}          # Wall-clock time limit
```
That's a bit tricky because all of strong-scaling experiments used high-memory nodes except for the 64-node experiment which uses the normal nodes. Initially, I wanted to use high-memory nodes in general but this turned out to be a problem as there are only 32 of them. However, as the problem size per node, decreases with increasing number of nodes in strong scaling, this was not a problem here but only for weak scaling. That is why the base problem size of weak scaling is only half the base problem size of strong scaling. I suggest to make a comment in the script.

mcw92 replied:
Update: I addressed this issue in the merged job script generation script.

mcw92 commented on scripts/experiments/generate_strong_scaling_job_scripts.py:
I merged the scripts as suggested.

fluegelk commented on scripts/experiments/generate_scaling_job_scripts.py:
```diff
@@ -0,0 +1,173 @@
+    submit: bool = False,
+) -> None:
+    """
+    Generate the job scripts for the weak scaling experiments.
```
```suggestion
    Generate the job scripts for the scaling experiments.
```

fluegelk commented on scripts/experiments/generate_scaling_job_scripts.py:
```diff
@@ -0,0 +1,173 @@
+        job_name = f"n{log_n_samples}_m{log_n_features}_{scaling_type}_{n_nodes}"
+        job_script_name = f"{job_name}.sh"
+        script_content = f"""#!/bin/bash
+#SBATCH --job-name={job_name}  # Job name
+#SBATCH --partition=cpuonly    # Queue for resource allocation
+#SBATCH --time={time}          # Wall-clock time limit
+#SBATCH --mem={mem}mb                # Main memory (use standard nodes)
+#SBATCH --cpus-per-task=76            # Number of CPUs required per (MPI) task
```
```suggestion
#SBATCH --job-name={job_name}         # Job name
#SBATCH --partition=cpuonly           # Queue for resource allocation
#SBATCH --time={time}                 # Wall-clock time limit
#SBATCH --mem={mem}mb                 # Main memory (use standard nodes)
#SBATCH --cpus-per-task=76            # Number of CPUs required per (MPI) task
```

fluegelk reviewed (approved):
Looks good to me now, only two minor formatting/comment things left.
